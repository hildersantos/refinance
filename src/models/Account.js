import { types, getParent } from "mobx-state-tree";
import { values } from "mobx";
import { Transaction } from "./Transaction";

export const Account = types
  .model({
    id: types.identifier,
    name: types.string,
    initialBalance: 0,
    transactions: types.optional(types.array(Transaction), [])
  })
  .views(self => ({
    get balance() {
      return values(self.transactions).reduce((sum, t) => {
        return sum + t.value;
      }, 0);
    }
  }))
  .actions(self => ({
    createTransaction: newTransaction => {
      // Checo se Ã© uma despesa, em caso positivo, mudo o total...
      self.transactions.push(newTransaction);
      try {
        getParent(self, 2).addTransaction(newTransaction.id);
      } catch (error) {
        return;
      }
    }
  }));
