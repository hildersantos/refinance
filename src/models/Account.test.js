import { Account } from "./Account";
import { applySnapshot } from "mobx-state-tree";

const initialAccount = Account.create({
  name: "Inter"
});
Date.now = jest.fn(() => 1482363367071);

it("can create an account", () => {
  const account = initialAccount;

  expect(account.transactions.length).toBe(0);
});

it("can create a transaction into an account", () => {
  const account = initialAccount;

  account.createTransaction({
    type: "d",
    value: 1000,
    description: "Despesa",
    date: Date.now()
  });

  expect(account.transactions[0].toJSON()).toEqual({
    type: "d",
    value: 1000,
    description: "Despesa",
    date: Date.now()
  });
});

it("can get account balance", () => {
  const account = initialAccount;

  const snapshot = {
    name: "Inter",
    transactions: [
      {
        type: "d",
        value: 1000,
        description: "Despesa",
        date: Date.now()
      },
      {
        type: "r",
        value: 3000,
        description: "Receita",
        date: Date.now()
      }
    ]
  };

  applySnapshot(account, snapshot);

  expect(account.balance).toBe(2000);
});
