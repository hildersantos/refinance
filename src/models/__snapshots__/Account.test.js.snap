// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a transaction into an account 1`] = `
Object {
  "id": "a1",
  "initialBalance": 0,
  "name": "Inter",
  "transactions": Array [
    Object {
      "date": 1482363367071,
      "description": "Despesa",
      "id": "t1",
      "isPaid": false,
      "type": "d",
      "value": -1000,
    },
  ],
}
`;
